### PR TITLE
chore(core): type tag is a short everywhere in the codebase, plus add improvements on function parser

### PR DESCRIPTION
**Pull request by marregui**
- Type tag becomes a short everywhere in the codebase.
- Function parser gets a small review, improving constant resolution, and making it easier to debug.
- More tests, revisiting tests.
- Better error messages around implicit casts.
- Better understanding from me.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 475 / 488 (97.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Long128TypeConstant.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|171|180|95.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryDescriptor.java|59|60|98.33%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|170|171|99.42%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/DateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserSupport.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToIPv4FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/CompiledFilterIRSerializer.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndStringListFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig607.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/ParallelCsvFileImporter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/LongFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/RecordToRowCopierUtils.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 544 / 585 (92.99%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Long128TypeConstant.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|57|82|69.51%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|169|176|96.02%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryDescriptor.java|67|69|97.10%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|170|171|99.42%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/DateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserSupport.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToIPv4FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/CompiledFilterIRSerializer.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndStringListFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig607.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/ParallelCsvFileImporter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/LongFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/RecordToRowCopierUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 555 / 596 (93.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Long128TypeConstant.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|57|82|69.51%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|180|187|96.26%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryDescriptor.java|67|69|97.10%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|170|171|99.42%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/DateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserSupport.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToIPv4FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/CompiledFilterIRSerializer.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndStringListFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig607.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/ParallelCsvFileImporter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/LongFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/RecordToRowCopierUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 555 / 596 (93.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Long128TypeConstant.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|57|82|69.51%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|180|187|96.26%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryDescriptor.java|67|69|97.10%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|170|171|99.42%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/DateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserSupport.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToIPv4FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/CompiledFilterIRSerializer.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndStringListFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig607.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/ParallelCsvFileImporter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/LongFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/RecordToRowCopierUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 555 / 596 (93.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Long128TypeConstant.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|57|82|69.51%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|180|187|96.26%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryDescriptor.java|67|69|97.10%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|170|171|99.42%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/DateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserSupport.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToIPv4FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/CompiledFilterIRSerializer.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndStringListFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig607.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/ParallelCsvFileImporter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/LongFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/RecordToRowCopierUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|1|1|100.00%|

**Review by amunra — DISMISSED** [review]
Lookin' good!

I spent some time w/ Miguel over Zoom and took the liberty to clean up some code that was deeply nested.

The code in the function factory could benefit from a second pair of eyes, possibly Bolek's.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 558 / 600 (93.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Long128TypeConstant.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|57|82|69.51%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|185|193|95.85%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryDescriptor.java|67|69|97.10%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|170|171|99.42%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/DateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserSupport.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToIPv4FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/CompiledFilterIRSerializer.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndStringListFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig607.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/ParallelCsvFileImporter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/LongFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/RecordToRowCopierUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|1|1|100.00%|

**Comment by bluestreak01** [comment]
well, i did no have a chance to comprehend this PR earlier. It was the original idea that column type is `int`, 32bit and it is divided in two 16 bit parts, tag and modifier. 16 bit tag is future proofing, such as for arrays. Those could be implemented as own type ids, which doubles the types we have now. There could be other even like that so we could run out of `byte`.

If `byte` does not solve user pain, i would keep this as column types as `short`.

**Comment by marregui** [comment]
> well, i did no have a chance to comprehend this PR earlier. It was the original idea that column type is `int`, 32bit and it is divided in two 16 bit parts, tag and modifier. 16 bit tag is future proofing, such as for arrays. Those could be implemented as own type ids, which doubles the types we have now. There could be other even like that so we could run out of `byte`.
> 
> If `byte` does not solve user pain, i would keep this as column types as `short`.

I think I confused myself. Many months ago, when we were coding geohash as a type, I introduced this comment:

```
// ColumnType layout - 32bit
//
// | PGWire format | Extra type information | Type discriminant (tag) |
// +---------------+------------------------+-------------------------+
// |    1 bit      |        23 bits         |         8 bits          |
// +---------------+------------------------+-------------------------+
```

which is wrong. I then saw the biggest value for the type (NULL) and saw that was well below the limit of a byte, so I thought there was an error somewhere, and I made tag a byte. Now it is back to being a short. Apologies for the confusion.

amended in 013bd25d767c437719cd78576855bec6482d7b5f

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 565 / 607 (93.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Long128TypeConstant.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|57|82|69.51%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|182|190|95.79%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryDescriptor.java|66|68|97.06%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|168|169|99.41%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/DateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserSupport.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToIPv4FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/jit/CompiledFilterIRSerializer.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndStringListFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig607.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/ParallelCsvFileImporter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpParserImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/LongFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/RecordToRowCopierUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|1|1|100.00%|

File: core/src/test/java/io/questdb/test/cairo/ColumnTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+import static io.questdb.cairo.ColumnType.*;
+
+public class ColumnTypeTest {
+
+    @Test
+    public void testIsBuiltInWideningCast() {
+        for (byte type = BOOLEAN; type < NULL; type++) {
```
Comment (eugenels) [inline core/src/test/java/io/questdb/test/cairo/ColumnTypeTest.java:42]: short type?

File: core/src/test/java/io/questdb/test/cairo/ColumnTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+            Assert.assertTrue(isBuiltInWideningCast(NULL, type));
+        }
+        Assert.assertTrue(isBuiltInWideningCast(CHAR, SHORT));
+        Assert.assertTrue(isBuiltInWideningCast(TIMESTAMP, LONG));
+        for (byte fromType = BYTE; fromType < STRING; fromType++) {
```
Comment (eugenels) [inline core/src/test/java/io/questdb/test/cairo/ColumnTypeTest.java:47]: short fromType

File: core/src/test/java/io/questdb/test/cairo/ColumnTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                        "unknown      | 0         | 0       | 0    | 0     | 0    | 0   | 0    | 0    | 0         | 0     | 0      | 0      | 0      | 0       | 0       | 0        | 0      | 0       | 0      | 0    | 0      | 0      | 0      | 0       | 0       | 0    | 0        | 0            | 0      | 0    | 0       | 0       | \n" +
+                        "unknown      | 0         | 0       | 0    | 0     | 0    | 0   | 0    | 0    | 0         | 0     | 0      | 0      | 0      | 0       | 0       | 0        | 0      | 0       | 0      | 0    | 0      | 0      | 0      | 0       | 0       | 0    | 0        | 0            | 0      | 0    | 0       | 0       | \n" +
+                        "================================================================================================================================================================================================================================================================================================================= | \n",
+                matrix.toString()
+        );
+    }
+
+    @Test
+    public void testSizeOf() {
+        for (int type = UNDEFINED; type <= NULL; type++) {
```
Comment (eugenels) [inline core/src/test/java/io/questdb/test/cairo/ColumnTypeTest.java:138]: typeTab type should be consistent (short)

File: core/src/main/java/io/questdb/griffin/FunctionParser.java [hunk 4]
```diff
@@ -76,17 +73,11 @@ public FunctionParser(CairoConfiguration configuration, FunctionFactoryCache fun
...
     @NotNull
-    public static ScalarFunction createColumn(
-            int position,
-            CharSequence name,
-            RecordMetadata metadata
-    ) throws SqlException {
-        final int index = metadata.getColumnIndexQuiet(name);
-
+    public static ScalarFunction createColumn(int position, CharSequence name, RecordMetadata metadata) throws SqlException {
+        int index = metadata.getColumnIndexQuiet(name);
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/FunctionParser.java:77]: final ?

File: core/src/main/java/io/questdb/griffin/FunctionParser.java [hunk 5]
```diff
@@ -803,18 +804,24 @@ private Function createFunction(
...
+        switch (ColumnType.tagOf(fromType)) {
             case ColumnType.STRING:
             case ColumnType.SYMBOL:
-                if (toType == ColumnType.UUID) {
-                    return new CastStrToUuidFunctionFactory.Func(function);
-                }
                 if (toType == ColumnType.TIMESTAMP) {
                     return new CastStrToTimestampFunctionFactory.Func(function);
                 }
+                if (toType == ColumnType.LONG256) {
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/FunctionParser.java:813]: ColumnType.tagOf(toType) == ?

File: core/src/main/java/io/questdb/cairo/ColumnType.java [hunk 6]
```diff
@@ -24,25 +24,22 @@
...
-// +---------------+------------------------+-------------------------+
-// |    1 bit      |        23 bits         |         8 bits          |
-// +---------------+------------------------+-------------------------+
-
-/**
- * Column types as numeric (integer) values
- */
+// | PGWire format |  Extra type information | Tag               |
+// +---------------+-------------------------+-------------------+
+// |    1 bit      |         23 bits         |   8 bits          |
```
Comment (eugenels) [inline core/src/main/java/io/questdb/cairo/ColumnType.java:34]: comment is not in sync with implementation.
```
    private static final int TAG_SIZE = Short.SIZE;
    private static final int GEOBITS_SIZE = Byte.SIZE;
    private static final int TYPE_FLAG_DESIGNATED_TIMESTAMP = (1 << (1 + TAG_SIZE + GEOBITS_SIZE));
    private static final int TYPE_FLAG_GEOHASH = (1 << (TAG_SIZE + GEOBITS_SIZE));
    ```

File: core/src/main/java/io/questdb/cairo/ColumnType.java [hunk 7]
```diff
@@ -132,7 +126,7 @@ public static boolean isBinary(int columnType) {
     }
 
     public static boolean isBoolean(int columnType) {
-        return columnType == ColumnType.BOOLEAN;
+        return columnType == BOOLEAN;
```
Comment (eugenels) [inline core/src/main/java/io/questdb/cairo/ColumnType.java:129]: we should compare tags here: ColumnType.tagOf(columnType) ==
this applies to other comparison fns

File: core/src/main/java/io/questdb/cairo/ColumnType.java [hunk 8]
```diff
@@ -203,124 +203,147 @@ public static boolean isVariableLength(int columnType) {
...
         short tag = tagOf(columnType);
-        if (tag < TYPE_SIZE.length) {
-            return TYPE_SIZE[tag];
-        }
-        return -1;
+        return tag < TYPE_SIZE.length ? TYPE_SIZE[tag] : NOT_STORED_TYPE_SIZE;
     }
 
     public static short tagOf(int type) {
         return (short) (type & 0xFF);
```
Comment (eugenels) [inline core/src/main/java/io/questdb/cairo/ColumnType.java:238]: the mask should be 16 bits: return (short) (type & 0xFFFF);

File: core/src/main/java/io/questdb/griffin/FunctionParser.java (hunk 4 above)
Comment (marregui) [reply core/src/main/java/io/questdb/griffin/FunctionParser.java:77]: final :D that is a cosmetic. All it does is to make you think that the code is immutable. I view it mostly as a waste of space unless the method is 45 cm in length (of my screen) and I have to scroll around to figure out if the variable is immutable.... I think this debate is equivalent to:

# Option A)

```
if (null == variable) {
}
```

# Option B)

```
if (variable == null) {
}
```

matter of preference.

**Pull request closed without merging**